File: books/physical-ai-humanoid-robotics/code-examples/python/foundations/hello_robot.py
#!/usr/bin/env python3
"""
Hello Robot - Basic ROS2 Node Example
Chapter: Introduction to Physical AI & Humanoid Robotics
"""

import rclpy
from rclpy.node import Node
from std_msgs.msg import String


class HelloRobotNode(Node):
    """
    A simple ROS2 node that publishes a "Hello Robot" message
    """
    def __init__(self):
        super().__init__('hello_robot_node')
        self.publisher = self.create_publisher(String, 'robot_greeting', 10)
        timer_period = 1  # seconds
        self.timer = self.create_timer(timer_period, self.timer_callback)
        self.i = 0

    def timer_callback(self):
        msg = String()
        msg.data = f'Hello Robot! Message #{self.i}'
        self.publisher.publish(msg)
        self.get_logger().info(f'Publishing: "{msg.data}"')
        self.i += 1


def main(args=None):
    rclpy.init(args=args)
    hello_robot_node = HelloRobotNode()

    try:
        rclpy.spin(hello_robot_node)
    except KeyboardInterrupt:
        pass
    finally:
        hello_robot_node.destroy_node()
        rclpy.shutdown()


if __name__ == '__main__':
    main()

File: books/physical-ai-humanoid-robotics/code-examples/cpp/foundations/hello_robot.cpp
#include <iostream>
#include <memory>
#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/string.hpp"

using namespace std::chrono_literals;

class HelloRobotNode : public rclcpp::Node
{
public:
    HelloRobotNode() : Node("hello_robot_node")
    {
        publisher_ = this->create_publisher<std_msgs::msg::String>("robot_greeting", 10);
        timer_ = this->create_wall_timer(
            1000ms, std::bind(&HelloRobotNode::timer_callback, this));
        count_ = 0;
    }

private:
    void timer_callback()
    {
        auto message = std_msgs::msg::String();
        message.data = "Hello Robot! Message #" + std::to_string(count_++);
        RCLCPP_INFO(this->get_logger(), "Publishing: '%s'", message.data.c_str());
        publisher_->publish(message);
    }
    rclcpp::TimerBase::SharedPtr timer_;
    rclcpp::Publisher<std_msgs::msg::String>::SharedPtr publisher_;
    size_t count_;
};

int main(int argc, char * argv[])
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<HelloRobotNode>());
    rclcpp::shutdown();
    return 0;
}